### PR TITLE
Introduce session cache

### PR DESCRIPTION
**Pull request by zhangrui1203**
## Motivation

SGLang provides session-level KV cache management, enabling user-defined business logic to explicitly control advanced features such as the lifecycle of the KV cache, hierarchical caching, and replica count management. In large-scale LLM service scenarios within data centers, SGLang serves as a high-performance data processing node, better supporting business logic and achieving superior results in performance, cost efficiency, and user experience.

## Modifications

- implement session cache framework
- improve hicache local file to support session cache
- support GD2FS as session cache storage backend
- improve performance for get_dummy_flat_data_page
- bug fix for cache_finished_req

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/managers/io_struct.py [hunk 1]
```diff
@@ -143,6 +144,21 @@ class SessionParams:
     offset: Optional[int] = None
     replace: Optional[bool] = None
     drop_previous_output: Optional[bool] = None
+    # The session cache segments.
+    # Each segment is a dict with 'token_start', 'token_length',
+    # 'kv_uri', 'kv_start', and 'kv_length' keys.
+    old_kv_cache: Optional[List[Dict]] = None
+    new_kv_cache: Optional[List[Dict]] = None
```
Comment (xiezhq-hermann) [inline python/sglang/srt/managers/io_struct.py]: I suggest to use a different pair of names here, making `old_kv_cache` to be `stored_kv_cache`. We can keep `new_kv_cache`, or use `fresh_kv_cache`. The current old vs new is a bit confusing.

File: python/sglang/srt/managers/io_struct.py [hunk 2]
```diff
@@ -143,6 +144,21 @@ class SessionParams:
...
+
+    def __post_init__(self):
+        old = SessionCache(self.old_kv_cache)
+        new = SessionCache(self.new_kv_cache)
+        if len(old) > 0 and len(new) > 0 and old.token_end != new.token_start:
```
Comment (xiezhq-hermann) [inline python/sglang/srt/managers/io_struct.py:156]: this is a very strong assumption of execution, do you have following plans on fault tolerance?

File: python/sglang/srt/managers/schedule_batch.py [hunk 3]
```diff
@@ -1003,7 +1009,7 @@ def _check_str_based_finish(self):
             if len(self.sampling_params.stop_regex_strs) > 0:
                 for stop_regex_str in self.sampling_params.stop_regex_strs:
                     if re.search(stop_regex_str, tail_str):
-                        self.finished_reason = FINISHED_MATCHED_REGEX(
+                        self.finished_reason = FINISH_MATCHED_REGEX(
```
Comment (xiezhq-hermann) [inline python/sglang/srt/managers/schedule_batch.py]: let's have all sorts of style fix in a separate PR

File: python/sglang/srt/managers/scheduler.py [hunk 4]
```diff
@@ -1968,6 +1999,13 @@ def get_new_batch_prefill(self) -> Optional[ScheduleBatch]:
                 ):
                     break
 
+            if self.enable_session_cache:
+                session_id = req.session_id
+                if session_id is not None:
+                    done = self.tree_cache.check_session_prefetch_progress(session_id)
+                    if not done:
+                        continue
```
Comment (xiezhq-hermann) [inline python/sglang/srt/managers/scheduler.py:2052]: so here is a blocking call for prefetching for the session cache, will there be more policies for it?

File: python/sglang/srt/mem_cache/memory_pool.py [hunk 5]
```diff
@@ -809,6 +816,31 @@ def get_cpu_copy(self, indices):
...
+            )
+
+        kv_host = torch.zeros(
+            (2, self.layer_num, len(indices), self.head_num, self.head_dim),
+            dtype=self.store_dtype,
+            device="cpu",
+            pin_memory=True,
+        )
+
+        torch.cuda.synchronize()
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/memory_pool.py:880]: probably better use stream sync for lower overhead

File: python/sglang/srt/mem_cache/session_cache.py [hunk 6]
```diff
@@ -0,0 +1,615 @@
...
+
+            flat_data = mem_pool_host.get_dummy_flat_data_page(page_num)
+
+            storage, filepath = SessionCacheStorageManager.get_storage(
+                seg.kv_uri, storage_config, mem_pool_host
+            )
+
+            storage.load(filepath, seg.kv_start, flat_data)
+
+            mem_pool_host.set_from_flat_data(mem_indices, flat_data)
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/session_cache.py:509]: will there be a zero-copy mechanism for loading data from the storage?

File: python/sglang/srt/mem_cache/session_cache.py [hunk 7]
```diff
@@ -0,0 +1,615 @@
...
+
+        kv_cache = self._shift_real_token()
+        for seg in kv_cache:
+            seg_data = flat_data[:, :, seg.token_start : seg.token_end, :, :]
+
+            storage, filepath = SessionCacheStorageManager.get_storage(
+                seg.kv_uri, storage_config, mem_pool_host
+            )
+
+            storage.save(filepath, seg.kv_start, seg_data)
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/session_cache.py:543]: same comments for zero-copy

File: python/sglang/srt/mem_cache/radix_cache.py [hunk 8]
```diff
@@ -447,7 +453,7 @@ def cache_finished_req(self, req: Req, is_insert: bool = True):
         ]
 
         # Maybe convert to bigram keys for EAGLE
-        keys = convert_to_bigram_key(req.fill_ids) if self.is_eagle else req.fill_ids
+        keys = convert_to_bigram_key(token_ids) if self.is_eagle else token_ids
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/radix_cache.py:463]: not sure about this one, if it's more of a style change please make it in a separate PR, otherwise will need @ispobock's opinion on it

**Comment by xiezhq-hermann** [comment]
Thank you @zhangrui1203 @pizhenwei for the PR, overall a very neat PR and improved a lot since my initial prototype: https://github.com/sgl-project/sglang/tree/xiezhq-session. I left some comments and have a few addition questions:
1. I do think the session cache abstraction can be compatible with other storage backends with only minor interface changes, e.g., providing a file interface to load and save session specific KV caches. It's ok for now to only support your backend but please help think about it in future refactoring.
2. Zero-copy from storage to memory is still missing, leaving fairly substantial performance on the table.
3. Can you prepare a test case just like other storage backends do, that way we have a correctness guarantee. It would be better to have a simple performance benchmark too, we can adapt the existing multi-turn benchmark script for it a bit. A validation on the session abstraction achieving close or even better performance over existing pages would be great.

File: python/sglang/srt/managers/io_struct.py (hunk 1 above)
Comment (zhangrui1203) [reply python/sglang/srt/managers/io_struct.py]: OK

File: python/sglang/srt/managers/io_struct.py (hunk 2 above)
Comment (zhangrui1203) [reply python/sglang/srt/managers/io_struct.py:156]: If users intend to utilize the session cache (by specifying session_id along with stored_kv_cache and fresh_kv_cache), they must strictly adhere to the following requirements:

1. The internal segments of both stored_kv_cache and fresh_kv_cache must be contiguous in terms of sequence positions.
2. The fresh_kv_cache must be directly appendable to the tail of the stored_kv_cache, maintaining sequential continuity.

If these conditions are not satisfied, SGLang must return an error. This is because the session cache relies on metadata that is managed and persisted by the control plane (i.e., the user). As a data-plane component, SGLang cannot interpret or recover from malformed or inconsistent control-plane metadata.

File: python/sglang/srt/managers/schedule_batch.py (hunk 3 above)
Comment (zhangrui1203) [reply python/sglang/srt/managers/schedule_batch.py]: OK

File: python/sglang/srt/managers/scheduler.py (hunk 4 above)
Comment (zhangrui1203) [reply python/sglang/srt/managers/scheduler.py:2052]: Currently, a generate request will block indefinitely until the session cache prefetch operation completes. In the future, timeout support will be added to prevent the generate request from hanging due to excessively long prefetch operations; however, this capability depends on the implementation of the backend storage system.

File: python/sglang/srt/mem_cache/memory_pool.py (hunk 5 above)
Comment (zhangrui1203) [reply python/sglang/srt/mem_cache/memory_pool.py:880]: OK

File: python/sglang/srt/mem_cache/radix_cache.py (hunk 8 above)
Comment (zhangrui1203) [reply python/sglang/srt/mem_cache/radix_cache.py:463]: We noticed that `req.fill_ids` here only includes `req.origin_input_ids` and does **not** include `req.output_ids`, which causes the cached KV cache to be missing the KV entries corresponding to `req.output_ids`.

**Comment by zhangrui1203** [comment]
Hi, @xiezhq-hermann 
The main changes in the new version are as follows:

- rename `old_kv_cache`/`new_kv_cache` to `stored_kv_cache`/`fresh_kv_cache`
- remove style fix
- use `torch.cuda.stream()` instead of `torch.cuda.synchronize()` in `get_flat_data`
- rebase to the latest commit

The test case is currently being added.

**Comment by zhangrui1203** [comment]
@xiezhq-hermann 
One more thing: While adding accuracy tests, we noticed they depend on @sgl.function, which would require API changes. Are there simpler or more standard ways to test this?

File: benchmark/hicache/bench_multiturn.py [hunk 9]
```diff
@@ -356,6 +376,58 @@ def __init__(self, args):
...
+        if not self.enable_session_cache:
+            return
+
+        self.sessions: Dict[int, Any] = {
+            i: {"id": f"session_{i}"} for i in range(num_clients)
+        }
+
+        # Create session cache files
+        hicache_file_path = os.getenv(
+            "SGLANG_HICACHE_FILE_BACKEND_STORAGE_DIR", "/tmp/hicache"
```
Comment (alphabetc1) [inline benchmark/hicache/bench_multiturn.py]: nit: please put all environment variables in `environ.py` and use `envs.VAR_NAME.get()`

**Comment by zhangrui1203** [comment]
Hi, @xiezhq-hermann 
All CI runs have completed. The failures appear unrelated to this PR, and some may pass with a retry.

**Comment by xiezhq-hermann** [comment]
@zhangrui1203 thank you for all your effort. How has the accuracy testing been going? My understanding is that as long as we’re using the same session, the session cache should be in effect.

Regarding the merge, I’d suggest splitting it into multiple independent PRs. That should make the code review easier and keep each commit atomic.
